lib: allow incoming Initial packets lost/reordered

junhochoi opened the pull request:
Let's think the following scenario:

0. Client sends Initial pn=0
1. Server sends Initial pn=0 -> Lost
2. Server sends Handshake pn=0

Currently when Initial#0 is lost, Client received Handshake#0 and
close the connection immediately because now it gets `Error::CryptoFail`
because Handshake#0 cannot be decrypted.

This can happen when 1 is lost, or reordered so 2 is delivered before 1.
Server doesn't know the connection is closed in the Client,
so keep trying to retransmit and timed out eventually.

The fix is not to immediately close the connection for invalid
incoming packets when Client has no valid received packets. To allow
loss recovery from Server or reordered delivery, now it has a threshold
to allow -- `recovery::PACKET_THRESHOLD` constant is used here because
it's already used for loss recovery/reordering detection. Within
the threshold, it simply discard the packet, waiting Server to
retransmit unless reordered and Client will retransmit its Initial
until the correct packet is received.

Server behavior is not changed (close the connection when
invalid packet is received).

Test Update:
- invalid_initial() is renamed to invalid_initial_server().
- invalid_initial_client() is added to check the client behavior.

junhochoi commented:
When using qvalve, `< 1r drop` or `< 1r reor` can reproduce this.

junhochoi commented:
Note that this is a client side fix - the server is not impacted by this change I think.

ghedo commented:
> Note that this is a client side fix - the server is not impacted by this change I think.

The `invalid_initial` test is intended to test for the server behavior, so the above is not really true since the original test would fail with these changes. The server won't close the connection anymore, which kind of defeats the whole point of `drop_pkt_on_err()`.

I think a better fix would be to not do the `recv_count` check on the client-side at all, since the client already sent a packet then the idle timeout is already armed and the connection will time-out eventually.

ghedo commented:
Also should probably add a separate test specifically for the client behavior.

junhochoi commented:
> The `invalid_initial` test is intended to test for the server behavior, so the above is not really true since the original test would fail with these changes. The server won't close the connection anymore, which kind of defeats the whole point of `drop_pkt_on_err()`.

I bring back `invalid_packet()` test and renamed to `invalid_packet_server()` since it's server behavior test.
 
> I think a better fix would be to not do the `recv_count` check on the client-side at all, since the client already sent a packet then the idle timeout is already armed and the connection will time-out eventually.

Now I split the case using `is_server` flag -- existing behavior for server but client will ignore a first few invalid packets as described in the PR.

> Also should probably add a separate test specifically for the client behavior.

invalid_packet_client() to test the client behavior.

In summary, existing behavior (close the connection immediately when server received an invalid packet) is kept -- tested by invalid_packet_server() and client will have a new behavior (ignore a few packets in the beginning) -- tested by invalid_packet_client().

ghedo commented:
@junhochoi my point was that you should not check for `sent_count` at all. The client already has the idle timeout to make sure the connection is closed.

junhochoi commented:
> @junhochoi my point was that you should not check for `sent_count` at all. The client already has the idle timeout to make sure the connection is closed.

Idle timeout still kicks in when there is inactivity. Here I use sent_count to check if the client is still in the beginning and during that time it discards invalid packet instead of closing. After the threshold, client will close the connection by idle time out or invalid packets received as usual. recv_count can't be used because it doesn't change when receiving invalid packet.

Ideally it'd better if we can count ignored packets but it seems need more changes

ghedo commented:
So, maybe I wasn't clear. What I'm saying is that we should only change the check to:
```rust
     if is_server && recv_count == 0 {
     ...
     }
```
the rest is unnecessary and breaks the client-side idle timeout so it should be removed.

junhochoi commented:
> So, maybe I wasn't clear. What I'm saying is that we should only change the check to:
> 
> ```rust
>      if is_server && recv_count == 0 {
>      ...
>      }
> ```
> 
> the rest is unnecessary and breaks the client-side idle timeout (as it causes the connection to be closed before the timeout expires) so it should be removed.

I see. With the changes above now it becomes as follows:

- Client: always ignore invalid packet incoming, relying on the timeout.
- Server: same as before (invalid initial packet before receiving anything will close the connection)

Is it something you want? I'm ok with it. If so I will make changes accordingly including tests.

ghedo commented on src/recovery.rs:
```diff
@@ -43,7 +43,7 @@ use crate::packet;
 
 // Loss Recovery
-const PACKET_THRESHOLD: u64 = 3;
+pub const PACKET_THRESHOLD: u64 = 3;
```
```suggestion
const PACKET_THRESHOLD: u64 = 3;
```

ghedo commented on src/lib.rs:
```diff
@@ -5285,6 +5302,41 @@ mod tests {
+        assert_eq!(pipe.client.recv(&mut buf[..written]), Ok(68));
+
+        // The connection should be alive.
+        assert_eq!(pipe.client.is_closed(), false);
```
We should also check that the idle timeout is actually set, just to make sure.
